src/main/java/Duke.java

```diff
@@ -1,10 +1,86 @@
+import java.io.FileNotFoundException;
+
+import utility.Choice;
+import utility.Parser;
+import utility.Storage;
+import utility.TaskList;
+import utility.Ui;
+
 public class Duke {
-    public static void main(String[] args) {
-        String logo = " ____        _        \n"
-                + "|  _ \\ _   _| | _____ \n"
-                + "| | | | | | | |/ / _ \\\n"
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+
+    /**
+     * Used for printing messages to enhance user experience.
+     */
+    private final Ui ui;
+
+    /**
+     * Controls the reading and writing of data to hard disk.
+     */
+    private final Storage storage;
+
+    /**
+     * Data structure for managing tasks.
+     */
+    private TaskList tasks;
+
+    /**
+     * Responsible for reading user input and parsing it.
+     */
+    private final Parser parser;
+
+    /**
+     * Creates Duke, a personal assistant bot.
+     */
+    Duke() {
+        ui = new Ui();
+        parser = new Parser();
+        storage = new Storage("data.txt");
+        ui.showWelcome();
+        try {
+            tasks = new TaskList(storage.load());
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * You should have your own function to generate a response to user input.
+     * Replace this stub with your completed method.
+     */
+    public String getResponse(String input) {
+        ui.showLine();
+        try {
+            Choice c = parser.parseCommand(input);
+            System.out.println(c);
+            switch (c) {
```

Comment:
Perhaps instead of using the full input for each task, you could consider using the parser to prepare the input accordingly? Cos I noticed you work with substrings within the individual task methods, causing the taskList to be quite lengthy! So yeah perhaps some of that data processing responsibility can be abstracted to the parser?